### ISSUE TITLE
ClassCastException in Chunk#map

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#864]
```scala
@ import $ivy.`co.fs2::fs2-core:0.9.4`, fs2._, scala.concurrent.duration._
import $ivy.$                       , fs2._, scala.concurrent.duration._
@ Chunk.longs(new Array[Long](32)).map(_.toDouble)
java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Long
  scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:105)
  scala.Function1.apply$mcJJ$sp(Function1.scala:36)
  fs2.Chunk$Longs.mapL(Chunk.scala:752)
  fs2.Chunk$Longs.map(Chunk.scala:719)
  $sess.cmd1$.<init>(cmd1.sc:1)
  $sess.cmd1$.<clinit>(cmd1.sc:-1)
```

@djspiewak any ideas?

**Comment by durban** [comment]
In 2.11 it seems to work fine. Maybe it's related to the different encoding of lambdas in 2.12? (I'm just guessing.)

As far as I understand, `fs2.Chunk$Longs.mapL` shouldn't even be called in this case (it's for `Long => Long`). The `LL` extractor uses `.getClass.getSuperclass` on a function object created from the lambda. In 2.12:

```scala
scala> val f: Long => Double = _.toDouble
f: Long => Double = $$Lambda$1840/1922630329@ed5bbf7

scala> f.getClass.getSuperclass
res6: Class[?0] forSome { type ?0 >: ?0; type ?0 <: Long => Double } = class java.lang.Object
```

(Note the `Object`.)

In 2.11:

```scala
scala> f.getClass.getSuperclass
res2: Class[?0] forSome { type ?0 >: ?0; type ?0 <: Long => Double } = class scala.runtime.AbstractFunction1$mcDJ$sp
```

**Comment by yilinwei** [comment]
@durban I think that you're right that it's the lambdas in Java 8; I've had similar problems when moving Java 7 code to Java 8 code.

I note that the comment:

```
// copy-pasted code below for each primitive
  // sadly, @specialized does not work here since the generated class names are
  // not human readable and we want to be able to use these type names in pattern
  // matching, e.g. `h.receive { case (bits: Booleans) #: h => /* do stuff unboxed */ } `
```

is against specialization of the `Chunk`, but I'm wondering whether it would be possible to do a mixture of `specialized` and extractors which give the same effects?

I know I'm not hugely familiar with the internal workings of fs2, but could something like [this](https://gist.github.com/yilinwei/ea93cc9d78dc7b372f6bfa2bf0d107f5) work?

```
scala> import fs2._

scala> Chunk2.Ints.unapply(Chunk2.Ints.empty)
res1: Option[fs2.Chunk2[Int]] = Some(fs2.Chunk2$Ints$$anon$2@305d8b69)

scala> Chunk2.Ints.unapply(Chunk2.Booleans.empty)
res2: Option[fs2.Chunk2[Int]] = None
```

**Comment by djspiewak** [comment]
Sorry to have completely forgotten about this.  It's *definitely* related to the new lambda encoding.  In fact, I'm concerned that the new lambda encoding might defeat this trick altogether.  I'll see what I can do.

**Comment by mpilquist** [comment]
This is fixed in the 0.10 series.
